src/NServiceBus.TransactionalSession/TransactionalSessionBase.cs

```diff
@@ -0,0 +1,132 @@
+namespace NServiceBus.TransactionalSession
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Extensibility;
+    using Persistence;
+    using Transport;
+
+    abstract class TransactionalSessionBase : ITransactionalSession
+    {
+        protected TransactionalSessionBase(
+            ICompletableSynchronizedStorageSession synchronizedStorageSession,
+            IMessageSession messageSession,
+            IMessageDispatcher dispatcher)
+        {
+            this.synchronizedStorageSession = synchronizedStorageSession;
+            this.messageSession = messageSession;
+            this.dispatcher = dispatcher;
+            pendingOperations = new PendingTransportOperations();
+        }
+
+        public ISynchronizedStorageSession SynchronizedStorageSession
+        {
+            get
+            {
+                if (!IsOpen)
+                {
+                    throw new InvalidOperationException(
+                        "Before accessing the SynchronizedStorageSession, make sure to open the session by calling the `Open`-method.");
+                }
+
+                return synchronizedStorageSession;
+            }
+        }
+
+        public string SessionId => options?.SessionId;
+
+        protected ContextBag Context => options.Extensions;
+
+        protected bool IsOpen => options != null;
+
+        public abstract Task Commit(CancellationToken cancellationToken = default);
+
+        public virtual Task Open(OpenSessionOptions options = null, CancellationToken cancellationToken = default)
+        {
+            if (IsOpen)
+            {
+                throw new InvalidOperationException($"This session is already open. {nameof(ITransactionalSession)}.{nameof(ITransactionalSession.Open)} should only be called once.");
+            }
+
+            this.options = options ?? new OpenSessionOptions();
+            return Task.CompletedTask;
+        }
+
+        ContextBag ITransactionalSession.PersisterSpecificOptions { get; } = new ContextBag();
+
+        public async Task Send(object message, SendOptions sendOptions, CancellationToken cancellationToken = default)
+        {
+            if (!IsOpen)
+            {
+                throw new InvalidOperationException("Before sending any messages, make sure to open the session by calling the `Open`-method.");
```

Comment:
This comment as a hidden dependency on the perister specific open methods e.g. OpenCosmosDB. That's why we say `Open`-method. Should we change this to say:

```suggestion
                throw new InvalidOperationException("The session has to be opened before sending any messages");
```